File: python/src/main/python/Q020.py
from src.main.python.Solution import Solution

# Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
#
# The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.


class Q020(Solution):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        if s:
            for char in s:
                if char == '(' or char == '[' or char == '{':
                    stack.append(char)
                else:
                    if char == ')' and (not stack or stack[-1] != '('):
                        return False
                    elif char == ']' and (not stack or stack[-1] != '['):
                        return False
                    elif char == '}' and (not stack or stack[-1] != '{'):
                        return False
                    else:
                        stack.pop()
        return not stack

File: java/src/main/java/solutions/Q020.java
package solutions;

import java.util.Stack;

/**
 * Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
 *
 * The brackets must close in the correct order, "()" and "()[]{}" are all valid but "(]" and "([)]" are not.
 */

public class Q020 extends Solution {

    public boolean isValid(String s) {
        if (s == null || s.length() < 2) {
            return false;
        } else {
            Stack<Character> stack = new Stack<>();
            for (char c : s.toCharArray()) {
                if (c == '(' || c == '[' || c == '{') {
                    stack.push(c);
                } else if (stack.empty()) {
                    return false;
                } else if (c == ')' && stack.pop() != '(') {
                    return false;
                } else if (c == ']' && stack.pop() != '[') {
                    return false;
                } else if (c == '}' && stack.pop() != '{') {
                    return false;
                }
            }
            return stack.empty();
        }
    }

}
